core/invoker/src/main/scala/whisk/core/containerpool/kubernetes/KubernetesContainer.scala

```diff
@@ -0,0 +1,159 @@
+   * @param labels labels to set on the container
+   * @param name optional name for the container
+   * @return a Future which either completes with a KubernetesContainer or one of two specific failures
+   */
+  def create(transid: TransactionId,
+             image: String,
+             userProvidedImage: Boolean = false,
+             environment: Map[String, String] = Map(),
+             labels: Map[String, String] = Map(),
+             name: Option[String] = None)(implicit kubernetes: KubernetesApi,
+                                          ec: ExecutionContext,
+                                          log: Logging): Future[KubernetesContainer] = {
+    implicit val tid = transid
+
+    val podName = name.getOrElse("").replace("_", "-").replaceAll("[()]", "").toLowerCase()
+    for {
+      id <- kubernetes.run(image, podName, environment, labels).recoverWith {
+        case _ => Future.failed(WhiskContainerStartupError(s"Failed to run container with image '${image}'."))
+      }
+      ip <- kubernetes.inspectIPAddress(id).recoverWith {
+        // remove the container immediately if inspect failed as
+        // we cannot recover that case automatically
+        case _ =>
+          kubernetes.rm(id)
+          Future.failed(WhiskContainerStartupError(s"Failed to obtain IP address of container '${id.asString}'."))
+      }
+    } yield new KubernetesContainer(id, ip)
+  }
+}
+
+/**
+ * Represents a container as run by kubernetes.
+ *
+ * This class contains OpenWhisk specific behavior and as such does not necessarily
+ * use kubernetes commands to achieve the effects needed.
+ *
+ * @constructor
+ * @param id the id of the container
+ * @param addr the ip & port of the container
+ */
+class KubernetesContainer(protected val id: ContainerId, protected val addr: ContainerAddress)(
+  implicit kubernetes: KubernetesApi,
+  protected val ec: ExecutionContext,
+  protected val logging: Logging)
+    extends Container {
+
+  /** The last read timestamp in the log file */
+  private val lastTimestamp = new AtomicReference("")
+
+  protected val logsRetryCount = 15
+  protected val logsRetryWait = 100.millis
+
+  // no-op under Kubernetes
+  def suspend()(implicit transid: TransactionId): Future[Unit] = Future.successful({})
+
+  // no-op under Kubernetes
+  def resume()(implicit transid: TransactionId): Future[Unit] = Future.successful({})
+
+  override def destroy()(implicit transid: TransactionId): Future[Unit] = {
+    super.destroy()
+    kubernetes.rm(id)
+  }
+
+  def logs(limit: ByteSize, waitForSentinel: Boolean)(implicit transid: TransactionId): Source[ByteString, Any] = {
+
+    val activationMarkerCheck: ByteString ⇒ Boolean = { bs ⇒
+      if (bs.containsSlice(DockerContainer.ActivationSentinel)) {
+
+        val lastTS = bs.utf8String.lines.toSeq.lastOption
+          .getOrElse("""{"time":""}""")
+          .parseJson
+          .asJsObject
+          .fields("time")
+          .convertTo[String]
```

Comment:
It's not necessary to match on lines and non-emptyness here, since we have successfully passed the Framing-delimiter we know we only have 1 line at hand.

What about:

```scala
lastTimestamp.set(bs.utf8String.convertTo[LogLine].time)
```

Parsing to `LogLine` should be safe since you created the objects accordingly above.

Reply:
Very cool!